num_enum_derive/src/lib.rs

```diff
@@ -32,14 +33,32 @@ fn literal(i: i128) -> Expr {
     }
 }
 
-fn expr_to_int(val_exp: &Expr) -> Result<i128> {
-    Ok(match val_exp {
-        Expr::Lit(ref val_exp_lit) => match val_exp_lit.lit {
-            Lit::Int(ref lit_int) => lit_int.base10_parse()?,
-            _ => die!(val_exp => "Expected integer"),
-        },
-        _ => die!(val_exp => "Expected literal"),
-    })
+enum DiscriminantValue {
+    Literal(i128),
+    Expr(Expr),
+}
+
+fn parse_discriminant(val_exp: &Expr) -> Result<DiscriminantValue> {
+    match val_exp {
+        Expr::Lit(ref val_exp_lit) => {
+            if let Lit::Int(ref lit_int) = val_exp_lit.lit {
+                return Ok(DiscriminantValue::Literal(lit_int.base10_parse()?));
+            }
+        }
+        Expr::Unary(ExprUnary {
+            op: UnOp::Neg(..),
+            expr,
+            ..
+        }) => {
+            if let Expr::Lit(ref val_exp_lit) = **expr {
+                if let Lit::Int(ref lit_int) = val_exp_lit.lit {
+                    return Ok(DiscriminantValue::Literal(-lit_int.base10_parse()?));
+                }
+            }
```

Comment:
To avoid a bit of code duplication, maybe we could do something like:

```rs
    let mut sign = 1;
    let mut unsigned_expr = val_expr;
    if let Expr::Unary(ExprUnary { op: UnOp::Neg(..), expr }) = val_exp {
        unsigned_expr = &**expr;
        sign = -1;
    }
    match unsigned_expr {
        Expr::Lit(ref val_exp_lit) => match val_exp_lit.lit {
            Lit::Int(ref lit_int) => return Ok(
                DiscriminantValue::Literal(sign * lit_int.base10_parse()?)
            ),
            _ => {},
        },
        _ => {},
    }
    Ok(DiscriminantValue::Expr(val_exp.clone()))
```

but your code may be clearer, so up to you 🙂